### PR TITLE
Extract user deletion into a new service

### PR DESCRIPTION
**Pull request by FCsongradi**
Pull request for [issue 915](https://github.com/apigee/apigee-edge-drupal/issues/915).

### CODE REVIEW

File: src/UserRemovalHandler.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * Default user removal handler service implementation.
+ *
+ * User removal can happen programmatically and when that happens Drupal
+ * developers may not always want to remove developer accounts from Apigee
+ * as that destroy their API keys too. This service enables developers to
+ * use the service decorator pattern to tweak if and how the developer
+ * removal happens when a user is deleted.
+ */
+class UserRemovalHandler implements UserRemovalHandlerInterface {
```
Comment (mxr576) [inline src/UserRemovalHandler.php:35]: Suggedted name `RemoveDeveloperWithUserUserRemovalHandler`.

Make  the class `final` mark the `entityTypeManager` property as `private`, and use native typehints.

Comment (mxr576) [inline src/UserRemovalHandler.php]:
```suggestion
final class RemoveDeveloperWithUserUserRemovalHandler implements UserRemovalHandlerInterface {
```

File: src/UserRemovalHandler.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+  /**
+   * Entity type manager service.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
```
Comment (mxr576) [inline src/UserRemovalHandler.php]:
```suggestion
  private EntityTypeManagerInterface $entityTypeManager;
```

File: src/UserRemovalHandler.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+  /**
+   * UserRemovalHandler constructor.
```
Comment (mxr576) [inline src/UserRemovalHandler.php]:
```suggestion
   * RemoveDeveloperWithUserUserRemovalHandler constructor.
```

File: src/UserRemovalHandler.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+   *    Entity type manager service.
+   */
+  public function __construct(EntityTypeManagerInterface $entity_type_manager) {
+    $this->entityTypeManager = $entity_type_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function userDelete(UserInterface $account): void {
```
Comment (mxr576) [inline src/UserRemovalHandler.php]:
```suggestion
  public function __invoke(UserInterface $account): void {
```

File: src/UserRemovalHandler.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    if ($account->isAnonymous()) {
+      return;
+    }
+
+    try {
+      /** @var \Drupal\apigee_edge\Entity\Developer $developer */
+      $developer = $this->entityTypeManager->getStorage('developer')->load($account->getEmail());
+      // Sanity check, the developer may not exist in Apigee Edge.
+      if ($developer) {
+        $developer->delete();
```
Comment (mxr576) [inline src/UserRemovalHandler.php]: Would be nice to have an `info` level log message in logs about the removal of the developer. I know this is a change compared with the original implementation, but only a minor one.

File: src/UserRemovalHandlerInterface.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+  /**
+   * Deletion of developer.
+   *
+   * @param \Drupal\user\UserInterface $account
+   *   The Drupal user entity.
+   *
+   * @return void
+   */
+  public function userDelete(UserInterface $account): void;
```
Comment (mxr576) [inline src/UserRemovalHandlerInterface.php]:
```suggestion
  public function __invoke(UserInterface $account): void;
```

File: apigee_edge.module [hunk 7]
```diff
@@ -1486,27 +1487,7 @@ function apigee_edge_user_cancel_methods_alter(&$methods) {
...
-    }
-  }
-  catch (\Exception $exception) {
-    $context = [
-      '@developer' => $account->getEmail(),
-      '@message' => (string) $exception,
-    ];
-    watchdog_exception('apigee_edge', $exception, 'Could not delete @developer developer entity. @message %function (line %line of %file). <pre>@backtrace_string</pre>', $context);
-  }
+  \Drupal::classResolver(UserRemovalHandler::class)->userDelete($account);
```
Comment (mxr576) [inline apigee_edge.module]:
```suggestion
  (\Drupal::service('apigee_edge.user_removal_handler'))($account);
```

File: apigee_edge.module [hunk 8]
```diff
@@ -38,6 +38,7 @@ use Drupal\apigee_edge\Form\DeveloperSettingsForm;
 use Drupal\apigee_edge\JobExecutor;
 use Drupal\apigee_edge\Plugin\Field\FieldType\ApigeeEdgeDeveloperIdFieldItem;
 use Drupal\apigee_edge\Plugin\Validation\Constraint\DeveloperEmailUniqueValidator;
+use Drupal\apigee_edge\UserRemovalHandler;
```
Comment (mxr576) [inline apigee_edge.module]: Maybe place this class and the interface under the `Drupal\apigee_edge\User` namespace

File: src/UserRemovalHandler.php (hunk 5 above)
Comment (FCsongradi) [reply src/UserRemovalHandler.php]: Added info level logging when the developer is deleted.

File: apigee_edge.module (hunk 8 above)
Comment (FCsongradi) [reply apigee_edge.module]: Refactored

**Comment by mxr576** [comment]
This is a 2.x backport of #919.

**Review by mxr576 — APPROVED** [review]

**Comment by mxr576** [comment]
@FCsongradi #919 has changes, backporting them can wait until a review from maintainers.

**Comment by mxr576** [comment]
The build failed because Codecove update failed.

Should be ready for review.

**Comment by shishir-intelli** [comment]
> The build failed because Codecove update failed.
> 
> Should be ready for review.

@mxr576 
Lets rerun it, probably codecov uploading issue.

**Review by shishir-intelli — APPROVED** [review]
LGTM

**Pull request merged**